xc/device.ml

```diff
@@ -1430,6 +1430,102 @@ let can_surprise_remove ~xs (x: device) = Generic.can_surprise_remove ~xs x
 
 module Dm = struct
 
+  module QMP_Event = struct
+    open Qmp
+    let (pipe_r, pipe_w) = Unix.pipe ()
+
+    let (>>=) m f = match m with | Some x -> f x | None -> ()
+    let (>>|) m f = match m with | Some _ -> () | None -> f ()
+
+    module Lookup = struct
+      let ftod, dtoc = Hashtbl.create 16, Hashtbl.create 16
+      let add c domid =
+        Hashtbl.replace ftod (Qmp_protocol.to_fd c) domid;
+        Hashtbl.replace dtoc domid c
+      let remove c domid =
+        Hashtbl.remove ftod (Qmp_protocol.to_fd c);
+        Hashtbl.remove dtoc domid
+      let domid_of fd = try Some (Hashtbl.find ftod fd) with Not_found -> None
+      let channel_of domid = try Some (Hashtbl.find dtoc domid) with Not_found -> None
+    end
+
+    module Monitor = struct
+      module Epoll = Core.Linux_ext.Epoll
+      module Flags = Core.Linux_ext.Epoll.Flags
+      let create () = (Core.Std.Or_error.ok_exn Epoll.create) ~num_file_descrs:1001 ~max_ready_events:1
+      let wakeup () =  (* write single byte to wake up Monitor.wait *)
+        if Unix.write pipe_w " " 0 1 <> 1 then
+          debug "Pipe wrote error, failed to wake up qmp event thread"
```

Comment:
write*

Reply:
Excuse me, I don't understand this comment.

Reply:
Thanks, change `wrote` to `write`